cmd/opentelemetry/app/exporter/elasticsearchexporter/spanstore.go

```diff
@@ -0,0 +1,209 @@
+	serviceCache     cache.Cache
+	spanIndexName    indexNameProvider
+	serviceIndexName indexNameProvider
+	translator       *estranslator.Translator
+}
+
+// NewEsSpanWriter creates new instance ofr EsSpanWriter
+func NewEsSpanWriter(params config.Configuration, logger *zap.Logger) (*EsSpanWriter, error) {
+	client, err := esclient.GetElasticsearchClient(params, logger)
+	if err != nil {
+		return nil, err
+	}
+	tagsKeysAsFields, err := config.LoadTagsFromFile(params.Tags.File)
+	if err != nil {
+		return nil, err
+	}
+	return &EsSpanWriter{
+		client:           client,
+		spanIndexName:    newIndexNameProvider("jaeger-span", params.IndexPrefix, params.UseReadWriteAliases),
+		serviceIndexName: newIndexNameProvider("jaeger-service", params.IndexPrefix, params.UseReadWriteAliases),
+		translator:       estranslator.NewTranslator(params.Tags.AllAsFields, tagsKeysAsFields, params.GetTagDotReplacement()),
+		serviceCache: cache.NewLRUWithOptions(
+			100000,
+			&cache.Options{
+				TTL: time.Hour * 12,
+			},
+		),
+	}, nil
+}
+
+func newIndexNameProvider(index, prefix string, useAliases bool) indexNameProvider {
+	if prefix != "" {
+		prefix = prefix + "-"
+		index = prefix + index
+	}
+	index = index + "-"
+	if useAliases {
+		index = index + "write"
+	}
+	return indexNameProvider{
+		index:    index,
+		useAlias: useAliases,
+	}
+}
+
+type indexNameProvider struct {
+	index    string
+	useAlias bool
+}
+
+func (n indexNameProvider) get(date time.Time) string {
+	if n.useAlias {
+		return n.index
+	}
+	spanDate := date.UTC().Format("2006-01-02")
+	return n.index + spanDate
+}
+
+// CreateTemplates creates index templates.
+func (w *EsSpanWriter) CreateTemplates(spanTemplate, serviceTemplate string) error {
+	err := w.client.PuTemplate("jaeger-span", strings.NewReader(spanTemplate))
+	if err != nil {
+		return err
+	}
+	err = w.client.PuTemplate("jaeger-service", strings.NewReader(serviceTemplate))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO remove
+var counter int64
+
+// WriteTraces writes traces to the storage
+func (w *EsSpanWriter) WriteTraces(_ context.Context, traces pdata.Traces) (int, error) {
+	atomic.AddInt64(&counter, int64(traces.SpanCount()))
+	fmt.Printf("counter: %d\n", counter)
+	fmt.Printf("storing batch size: %d\n", traces.SpanCount())
+	spans, err := w.translator.ConvertSpans(traces)
+	if err != nil {
+		return traces.SpanCount(), consumererror.Permanent(err)
+	}
+	errors, err := w.writeSpans(spans)
+	if err != nil {
+		os.Exit(1)
+	}
+	return errors, err
+}
+
+func (w *EsSpanWriter) writeSpans(spans []*dbmodel.Span) (int, error) {
+	buffer := &bytes.Buffer{}
+	operationsToSpanIndices := make([]int, len(spans))
+	for i, span := range spans {
+		data, err := json.Marshal(span)
+		if err != nil {
+			// TODO continue with other spans
+			return len(spans), err
+		}
+		indexName := w.spanIndexName.get(model.EpochMicrosecondsAsTime(span.StartTime))
+		operationsToSpanIndices = append(operationsToSpanIndices, i)
```

Comment:
Why is this repeated twice? on line 150

Reply:
The array holds a mapping from bulk operation to a span index number. 

It will be used to construct the retry error of the bulk operation fails. It's there twice because service operation might also fail and it has a different bulk ID.